hardinfo/dmi_util.c

```diff
@@ -0,0 +1,152 @@
+    { "baseboard-serial-number", "id/board_serial" },
+    { "baseboard-asset-tag", "id/board_asset_tag" },
+    { "system-product-name", "id/product_name" },
+    { "system-manufacturer", "id/sys_vendor" },
+    { "system-serial-number", "id/product_serial" },
+    { "system-product-family", "id/product_family" },
+    { "system-version", "id/product_version" },
+    { "system-uuid", "product_uuid" },
+    { "chassis-type", "id/chassis_type" },
+    { "chassis-serial-number", "id/chassis_serial" },
+    { "chassis-manufacturer", "id/chassis_vendor" },
+    { "chassis-version", "id/chassis_version" },
+    { "chassis-asset-tag", "id/chassis_asset_tag" },
+    { NULL, NULL }
+  };
+  const gchar *dmi_root = dmi_sysfs_root();
+  gchar *full_path = NULL, *ret = NULL;
+  gboolean spawned;
+  gchar *out, *err;
+
+  int i = 0;
+
+  /* try sysfs first */
+  if (dmi_root) {
+    while (tab_dmi_sysfs[i].id != NULL) {
+      if (strcmp(id_str, tab_dmi_sysfs[i].id) == 0) {
+        full_path = g_strdup_printf("%s/%s", dmi_root, tab_dmi_sysfs[i].path);
+        if (g_file_get_contents(full_path, &ret, NULL, NULL) )
+          goto dmi_str_done;
+        }
+      i++;
+    }
+  }
+
+  /* try dmidecode, but may require root */
+  full_path = g_strconcat("dmidecode -s ", id_str, NULL);
+  spawned = g_spawn_command_line_sync(full_path,
+            &out, &err, &i, NULL);
+  if (spawned) {
+    if (i == 0)
+      ret = out;
+    else
+      g_free(out);
+    g_free(err);
+  }
+
+dmi_str_done:
+  if (ret != NULL) {
+    ret = strend(ret, '\n');
+    ret = g_strstrip(ret);
+    if (strlen(ret) == 0) {
+      g_free(ret);
+      ret = NULL;
+    }
+  }
+  g_free(full_path);
+  return ret;
+}
+
+char *dmi_chassis_type_str(int chassis_type, int with_val) {
+    static const char *types[] = {
+        N_("Invalid chassis type (0)"),
+        N_("Unknown chassis type"), /* 1 is "Other", but not helpful in HardInfo */
+        N_("Unknown chassis type"),
+        N_("Desktop"),
+        N_("Low-profile Desktop"),
+        N_("Pizza Box"),
+        N_("Mini Tower"),
+        N_("Tower"),
+        N_("Portable"),
+        N_("Laptop"),
+        N_("Notebook"),
+        N_("Handheld"),
+        N_("Docking Station"),
+        N_("All-in-one"),
+        N_("Subnotebook"),
+        N_("Space-saving"),
+        N_("Lunch Box"),
+        N_("Main Server Chassis"),
+        N_("Expansion Chassis"),
+        N_("Sub Chassis"),
+        N_("Bus Expansion Chassis"),
+        N_("Peripheral Chassis"),
+        N_("RAID Chassis"),
+        N_("Rack Mount Chassis"),
+        N_("Sealed-case PC"),
+    };
+
+    if (chassis_type <= 0) {
+        gchar *chassis = dmi_get_str("chassis-type");
+        chassis_type = atoi(chassis);
+        g_free(chassis);
+    }
+
+    if (chassis_type >= 0 && chassis_type < G_N_ELEMENTS(types)) {
+        if (with_val)
+            return g_strdup_printf("[%d] %s", chassis_type, _(types[chassis_type]));
+        else
```

Comment:
Nitpicking, but `else` is not required if the previous `if` statement ended with a `return`.